Data/ByteString/Short/Internal.hs

```diff
@@ -446,6 +554,781 @@ concat sbss =
+                     nl = l + 1
+  in create nl $ \mba -> do
+      writeWord8Array mba 0 c
+      copyByteArray (asBA sbs) 0 mba 1 l
+{-# INLINE cons #-}
+
+-- | /O(1)/ Extract the last element of a ShortByteString, which must be finite and non-empty.
+-- An exception will be thrown in the case of an empty ShortByteString.
+--
+-- @since 0.11.3.0
+last :: ShortByteString -> Word8
+last = \sbs -> case null sbs of
+  True -> error "empty ShortByteString"
+  False -> indexWord8Array (asBA sbs) (length sbs - 1)
+{-# INLINE last #-}
+
+-- | /O(n)/ Extract the elements after the head of a ShortByteString, which must be non-empty.
+-- An exception will be thrown in the case of an empty ShortByteString.
+--
+-- Note: copies the entire byte array
+--
+-- @since 0.11.3.0
+tail :: ShortByteString -> ShortByteString
+tail = \sbs -> 
+  let l = length sbs
+      nl = l - 1
+  in case null sbs of
+      True -> error "empty ShortByteString"
+      False -> create nl $ \mba -> copyByteArray (asBA sbs) 1 mba 0 nl
+{-# INLINE tail #-}
+
+-- | /O(1)/ Extract the first element of a ShortByteString, which must be non-empty.
+-- An exception will be thrown in the case of an empty ShortByteString.
+--
+-- @since 0.11.3.0
+head :: ShortByteString -> Word8
+head = \sbs -> case null sbs of
+  True -> error "empty ShortByteString"
+  False -> indexWord8Array (asBA sbs) 0
+{-# INLINE head #-}
+
+-- | /O(n)/ Return all the elements of a 'ShortByteString' except the last one.
+-- An exception will be thrown in the case of an empty ShortByteString.
+--
+-- Note: copies the entire byte array
+--
+-- @since 0.11.3.0
+init :: ShortByteString -> ShortByteString
+init = \sbs ->
+  let l = length sbs
+      nl = l - 1
+  in case null sbs of
+      True -> error "empty ShortByteString"
+      False -> create nl $ \mba -> copyByteArray (asBA sbs) 0 mba 0 nl
+{-# INLINE init #-}
+
+
+-- ---------------------------------------------------------------------
+-- Transformations
+
+-- | /O(n)/ 'map' @f xs@ is the ShortByteString obtained by applying @f@ to each
+-- element of @xs@.
+--
+-- @since 0.11.3.0
+map :: (Word8 -> Word8) -> ShortByteString -> ShortByteString
+map f = \sbs ->
+    let l = length sbs
+        ba = asBA sbs
+    in create l (\mba -> go ba mba 0 l)
+  where
+    go :: BA -> MBA s -> Int -> Int -> ST s ()
+    go !ba !mba !i !l
+      | i >= l = return ()
+      | otherwise = do
+          let w = indexWord8Array ba i
+          writeWord8Array mba i (f w)
+          go ba mba (i+1) l
+
+
+-- | /O(n)/ 'reverse' @xs@ efficiently returns the elements of @xs@ in reverse order.
+--
+-- @since 0.11.3.0
+reverse :: ShortByteString -> ShortByteString
+reverse = \sbs ->
+    let l = length sbs
+        ba = asBA sbs
+    in create l (\mba -> go ba mba 0 l)
+  where
+    go :: BA -> MBA s -> Int -> Int -> ST s ()
+    go !ba !mba !i !l
+      | i >= l = return ()
+      | otherwise = do
+          let w = indexWord8Array ba i
+          writeWord8Array mba (l - 1 - i) w
```

Comment:
This can be implemented more efficiently with [`byteSwap`](https://hackage.haskell.org/package/base-4.14.0.0/docs/GHC-Exts.html#v:byteSwap64-35-), but I'm not sure if it's worth a trouble.

Reply:
I played a little with it... and it's annoying because the offset for `indexWord64Array` uses 8byte offsets.

So if we have e.g. 9 bytes input:

```
input: [b1, b2, b3, b4, b5, b6, b7, b8, b9]
mut:   [m1, m2, m3, m4, m5, m6, m7, m8, m9]
```

`b1` is written to `m9` via word8, because it's the overhang.

`indexWord64Array` won't allow to index `b2-b9` now. Splitting the input bytestring for indexing will cause a copy, so isn't an option.

Reply:
Ah, there's `indexWord8ArrayAsWord64#`...

Reply:
`indexWord8ArrayAsWord64#` is only available for GHC 8.6.x upwards it seems.